readthedocs/api/v2/serializers.py

```diff
@@ -378,3 +382,53 @@ def get_username(self, obj):
             or obj.extra_data.get("login")
             # FIXME: which one is GitLab?
         )
+
+
+class NotificationAttachedToRelatedField(serializers.RelatedField):
+
+    """
+    Attached to related field for Notifications.
+
+    Used together with ``rest-framework-generic-relations`` to accept multiple object types on ``attached_to``.
+
+    See https://github.com/LilyFoote/rest-framework-generic-relations
+    """
+
+    default_error_messages = {
+        "required": _("This field is required."),
+        "does_not_exist": _("Object does not exist."),
+        "incorrect_type": _(
+            "Incorrect type. Expected URL string, received {data_type}."
+        ),
+    }
+
+    def to_representation(self, value):
+        return f"{self.queryset.model._meta.model_name}/{value.pk}"
+
+    def to_internal_value(self, data):
+        # TODO: handle exceptions
+        model, pk = data.strip("/").split("/")
+        if self.queryset.model._meta.model_name != model:
+            self.fail("incorrect_type")
+
+        try:
+            return self.queryset.get(pk=pk)
+        except (ObjectDoesNotExist, ValueError, TypeError):
+            self.fail("does_not_exist")
+
+
+class NotificationSerializer(serializers.ModelSerializer):
+    # Accept different object types (Project, Build, User, etc) depending on what the notification is attached to.
+    # The client has to send a value like "<model>/<pk>".
+    # Example: "build/3522" will attach the notification to the Build object with id 3522
```

Comment:
Does this mean to query notifications we'd use something like `/api/v2/notifications/?attached_to=project/123`?

For consistency of the API, we'd normally use `/api/v2/notifications/?project__pk=123`

I don't quite know if it's something we have control over though.

Reply:
This endpoint is on APIv2 and it's only used by the builders via `POST`, so we won't use for listing or similar.

This field, `attached_to` has a special syntax because the `Notification` is tied to a `ContentType`. The `attached_to` field communicates the resource _and_ the id of that resource in the same field with a special syntax: `<resource>/<id>`.